Add support for an extra set of kernel packages and group configured with NO_HZ_FULL

gratian opened the pull request:
In certain types of real-time applications it is desirable to reserve CPU cores for real-time threads and avoid interference from the kernel as much as possible. This includes things like the kernel scheduler tick which can create jitter spikes.

Provide packages for a kernel configured with NO_HZ_FULL. This allows users with strict RT requirements to install a kernel that will adaptively try to shutdown the scheduler thick whenever possible. This comes at the expense of some overhead in user <-> kernel transitions (i.e. syscalls, exceptions, and interrupts).

This necessitaded some refactoring of existing recipes which assumed only two possible kernel configurations: 'current' and 'next'

gratian committed:
linux-nilrt: Rename linux-nilrt-next to linux-nilrt-alternate

Currently linux-nilrt-next.inc is included and used by
linux-nilrt-next_git.bb to build the 'next' version of NILRT kernel. The
logic it provides has general applicability to building alternate NILRT
kernels that can be installed via the package feed. Rename it to reflect
its extended use.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
packagegroups: Relocate kernel package groups under recipes-kernel

In preparation for adding more/different kernel flavors relocate kernel
related package groups under recipes-core/packagegroups for improved
readability and maintainability. No functional changes.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
kernel-devsrc: Factor out common bits into an include file

In preparation to adding more kernel flavors in addition to the existing
'next' one, factor out the common bits in kernel-devsrc into an include
file that can be reused.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
kernel: Add NO_HZ_FULL kernel flavor packages and package group

In certain types of real-time applications it is desirable to reserve CPU
cores for real-time threads and avoid interference from the kernel as much
as possible. This includes things like the kernel scheduler tick which can
create jitter spikes.

Provide packages for a kernel configured with NO_HZ_FULL. This allows users
with strict RT requirements to install a kernel that will adaptively try to
shutdown the scheduler thick whenever possible. This comes at the expense
of some overhead in user <-> kernel transitions (i.e. syscalls,
exceptions, and interrupts).

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
virtual/kernel: Relocate git source/branch related information

The current implementation assumes all current kernel flavors are built
from the same source and as such the git repo and branch information is
stored in the common include file. To allow for future expansion and adding
more kernel flavors relocate this information to the corresponding recipe.
This allows for finer grain control on the version used for different NILRT
kernel flavors.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
linux-nilrt-next: Relocate kernel branch information

The current implementation assumes all 'next' kernel flavors are built from
the same source and as such the git repo and branch information is stored
in a common include file. To allow for future expansion and adding more
kernel flavors relocate the 'next' kernel information information to the
corresponding recipe. This allows for finer grain control on the version
used for different NILRT kernel flavors.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
linux-nilrt-debug: Relocate kernel source and branch information

The current implementation assumes all kernel flavors are built from the
same source and as such the git repo and branch information is stored in a
common include file. To allow for future expansion and adding more kernel
flavors relocate the 'debug' kernel information information to the
corresponding recipe. This allows for finer grain control on the version
used for different NILRT kernel flavors.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
linux-nilrt: Include kernel type information in the version extension

Include the kernel type in the kernel version extension to help
differentiate between different kernel flavors and avoid name collisions
when additional kernels are installed.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
linux-nilrt-debug: Update and fix dependency issues

There are several issues with the current implementation that went
unnoticed:

  - an unconditional RDEPENDS is used to add 'kernel-debug-devicetree';
    this breaks installation on platforms that don't use device trees.

  - the recipe predates the transition to the dkms workflow.

Make the dependency on 'kernel-debug-devicetree' conditional on the machine
and provide a kernel-devsrc-debug recipe to handle the pre/post-install
steps needed for dkms based module installation.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian commented:
Attn: @ni/rtos 
FYI: @jhershbe

jhershbe commented:
Looks good to me.

amstewart reviewed (changes requested):
`linux-nilrt-extra` kind of sounds like it installs non-essential kernel components. What about `linux-nilrt-alternate` instead?

amstewart commented on recipes-kernel/linux/linux-nilrt-next_git.bb:
```diff
@@ -1,6 +1,6 @@
 DESCRIPTION = "NILRT linux kernel next development build"
 
-require linux-nilrt-next.inc
+require linux-nilrt-extra.inc
```
linux-nilrt-extra kind of sounds like it installs non-essential kernel components. What about linux-nilrt-alternate instead?

gratian replied:
Sounds good. I'll  change the name on the next patch rev.

amstewart reviewed (changes requested):
Disregard the above comment. Let's try this again, github.

amstewart commented on recipes-kernel/linux/linux-nilrt.inc:
```diff
@@ -1,10 +1,18 @@
+
 # Set the version for traceability purposes
-LINUX_VERSION_EXTENSION = "$([ -z "${BUILDNAME}" ] || echo '-')${BUILDNAME}"
+LINUX_VERSION_EXTENSION = "$([ "${LINUX_KERNEL_TYPE}" != "standard" ] && echo '-${LINUX_KERNEL_TYPE}')$([ -z "${BUILDNAME}" ] || echo '-')${BUILDNAME}"
```
Does the `LINUX_KERNEL_TYPE` get set to `standard` in the "current" kernel recipe? I don't see that happening anywhere.

gratian replied:
meta-nilrt/recipes-kernel/linux/linux-nilrt.inc (which is used by all nilrt kernel recipes) includes openembedded-core/meta/recipes-kernel/linux/linux-yocto.inc which sets:

```
# The distro or local.conf should set this, but if nobody cares...
LINUX_KERNEL_TYPE ??= "standard"
```
We've been using this default for virtual/kernel package builds. There might be an argument to be made here about setting the (weak defined) default in distro or local.conf as the comment in that .inc file implies but we've been operating using the linux-yocto.inc default for a while so I didn't mess with it.

amstewart replied:
OK. That's sufficient, I think.

amstewart commented on recipes-kernel/packagegroups/packagegroup-ni-debug-kernel.bb:
```diff
@@ -5,12 +5,11 @@ inherit packagegroup
 
 RDEPENDS_${PN} = "\
 	kernel-debug \
-	kernel-debug-base \
-	kernel-debug-vmlinux \
-	kernel-debug-image \
```
Why are we dropping all these from the packagegroup? Are they deprecated?

gratian replied:
The packagegroup is over-specifying the dependencies. The dependency chain for _kernel-debug_ already includes the _kernel-debug-base_ dependency which in turn depends on the package providing _kernel-debug-image_. Here's the dependency chain reconstructed from .ipks I have built to test this:

```
  kernel-debug_4.14+git0+db94e1d70a-r0_x64.ipk
  Depends: kernel-debug-4.14.146-rt67-debug-dev-nilrt

	kernel-debug-4.14.146-rt67-debug-dev-nilrt_4.14+git0+db94e1d70a-r0_x64.ipk
	Depends: kernel-debug-image-4.14.146-rt67-debug-dev-nilrt
	Provides: kernel-debug-4.14.146-rt67-debug-dev-nilrt, kernel-debug-base

		kernel-debug-image-4.14.146-rt67-debug-dev-nilrt_4.14+git0+db94e1d70a-r0_x64.ipk
		Depends: kernel-debug-image-bzimage-4.14.146-rt67-debug-dev-nilrt
		Provides: kernel-debug-image

			kernel-debug-image-bzimage-4.14.146-rt67-debug-dev-nilrt_4.14+git0+db94e1d70a-r0_x64.ipk
			Provides: kernel-debug-image-bzimage

```
The removal of _kernel-debug-vmlinux_ is accidental and caused by the fact I was reusing the nohz/next package groups (which don't need/want the vmlinux binary). For the debug kernel the vmlinux package is needed bc. it contains the un-stripped kernel image with debug symbols. Thanks for catching this; will fix in V2.

gratian commented:
Changes since V1:

  * rebase on latest nilrt/master/sumo

  * rename linux-nilrt-extra.inc -> linux-nilrt-alternate.inc and update corresponding recipes

  * re-add kernel-debug-vmlinux to packagegroup-ni-debug-kernel.bb

amstewart commented:
@chaitu236 Any objections?

chaitu236 commented:
> @chaitu236 Any objections?

No looks good to me.
